micro-frontends-with-angular-and-stenciljs.md

```diff
@@ -0,0 +1,150 @@
+publishAs: lehmamic
+ignorePost: true
+---
+
+Micro frontends is an emerging frontend architecture pattern defined by [Martin Fowler](https://martinfowler.com/articles/micro-frontends.html). The pattern slices a web frontend into loosely coupled, scaleable, and independent deployable components.
+
+## From UI monoliths to micro frontends
+
+Traditionally, when building a microservice application, we'll end up in a UI monolith.
+
+![ui-monolith.png](https://cdn.hashnode.com/res/hashnode/image/upload/v1657876177570/E56kWsPFi.png align="center")
+
+This hurts because we are losing our benefit from the microservice architecture in the frontend. It might be a drawback which you purposely take into account and pay the price for it.
+
+But with the micro frontends architecture pattern, we have a tool in our hand to bring the microservice pattern into the frontend. Each microservice has its own frontend which gets composed together in a shell web application.
+
+![scs.png](https://cdn.hashnode.com/res/hashnode/image/upload/v1657876204859/2qYNvTWb2.png align="center")
+
+## Sharing the layout in self-contained systems 
+
+We are actually doing the [self-contained systems](https://scs-architecture.org/) architecture in my current project. This is a special variation of the micro frontend pattern. Self-contained systems consist of multiple scalable and independent deployable web applications which are preferably linked together via HTML.
+We were facing the need to share layout components such as header and footer. Or building a dashboard, which integrates widgets from several web applications. We don't want to create libraries or something like that, because it is coupling the frameworks and their versions to the applications. The micro frontend pattern is exactly what we need.
+
+The sounds great. Unfortunately, the technologies (i.e. the web frameworks) did not catch up yet. And because Angular, React and VueJs are not ready for this scenario, a lot of manual work is required to compose a micro frontend application.
+
+## Techniques to compose the UI with micro frontends
+
+So let's have a look at our approach for building micro frontends supporting our self-contained systems. We are using Angular in our web applications. I don't want to discuss the framework decision in this article. It was more or less given by the knowledge of the team.
+
+There are multiple ways how we can compose micro frontends together:
+
+* Integrate the HTML via iFrame. This is quite old school, but it works. iFrame integration sounds easy, but I spent a lot of time in the past on sizing and responsiveness working across the applications.
+
+* Load and integrate the HTML via vanilla javascript. This is very painful to get it right. HTML requires javascript files, styles etc. It is a lot of manual work to load everything and make it work.
```

Comment:
javascript --> JavaScript

Reply:
I hate that one, I'll always write it differently ^^